File: synth_ai/tracing_v3/hooks.py
"""Hook system for extending tracing functionality.

The hook system provides a flexible way to extend the tracing system without
modifying core code. Hooks can be registered at various points in the tracing
lifecycle to perform custom processing, validation, or enrichment.

Hook Points:
-----------
- session_start: Called when a new session begins
- session_end: Called when a session ends
- timestep_start: Called when a new timestep begins
- timestep_end: Called when a timestep completes
- event_recorded: Called after an event is recorded
- message_recorded: Called after a message is recorded
- before_save: Called before saving a session to database
- after_save: Called after successful save to database

Hook Design:
-----------
Hooks are designed to be:
1. **Non-blocking**: Hooks should execute quickly to avoid impacting performance
2. **Fault-tolerant**: Hook failures don't stop the tracing pipeline
3. **Prioritized**: Hooks execute in priority order (highest first)
4. **Async-aware**: Support both sync and async callbacks

Common Use Cases:
----------------
- Data validation before storage
- Metric calculation and aggregation
- External system notifications
- Data enrichment and transformation
- Custom filtering and sampling
"""

from __future__ import annotations

import asyncio
from collections.abc import Callable
from dataclasses import dataclass
from typing import Any

from .abstractions import BaseEvent


@dataclass
class Hook:
    """A hook that can be registered with the tracer.

    Attributes:
        name: Unique identifier for the hook
        callback: Function to call when hook is triggered. Can be sync or async.
        event_types: Optional list of event types to filter on (for event_recorded hook)
        priority: Execution priority (higher numbers run first)
        enabled: Whether the hook is currently active
    """

    name: str
    callback: Callable
    event_types: list[str] | None = None
    priority: int = 0
    enabled: bool = True


class HookManager:
    """Manages hooks for session tracing.

    The HookManager maintains collections of hooks for each hook point and
    handles their execution. It ensures hooks are called in priority order
    and handles both sync and async callbacks appropriately.

    Thread Safety:
        The HookManager is designed to be thread-safe for registration and
        execution. Multiple async tasks can trigger hooks concurrently.
    """

    def __init__(self):
        self.hooks: dict[str, list[Hook]] = {
            "session_start": [],
            "session_end": [],
            "timestep_start": [],
            "timestep_end": [],
            "event_recorded": [],
            "message_recorded": [],
            "before_save": [],
            "after_save": [],
        }

    def register(
        self,
        event: str,
        callback: Callable,
        name: str | None = None,
        priority: int = 0,
        event_types: list[str] | None = None,
    ) -> Hook:
        """Register a new hook.

        Args:
            event: Hook point name (e.g., 'session_start', 'event_recorded')
            callback: Function to call. Signature depends on hook point:
                     - session_start(session_id: str, metadata: Dict)
                     - event_recorded(event_obj: BaseEvent)
                     - etc. (see individual hook point docs)
            name: Optional name for the hook (defaults to callback.__name__)
            priority: Execution priority (higher = earlier execution)
            event_types: For 'event_recorded' hook, filter to specific event types

        Returns:
            The created Hook instance

        Raises:
            ValueError: If the event name is not a valid hook point
        """
        if event not in self.hooks:
            raise ValueError(f"Unknown hook event: {event}")

        hook = Hook(
            name=name or getattr(callback, "__name__", "unknown"),
            callback=callback,
            event_types=event_types,
            priority=priority,
        )

        self.hooks[event].append(hook)
        self.hooks[event].sort(key=lambda h: h.priority, reverse=True)

        return hook

    def unregister(self, event: str, name: str):
        """Unregister a hook by name.

        Args:
            event: Hook point name
            name: Name of the hook to remove
        """
        if event not in self.hooks:
            return

        self.hooks[event] = [h for h in self.hooks[event] if h.name != name]

    async def trigger(self, event: str, *args, **kwargs) -> list[Any]:
        """Trigger all hooks for an event.

        Executes all registered hooks for the given event in priority order.
        Handles both sync and async callbacks appropriately. Exceptions in
        hooks are caught and logged but don't stop execution of other hooks.

        Args:
            event: Hook point name
            *args: Positional arguments passed to hook callbacks
            **kwargs: Keyword arguments passed to hook callbacks

        Returns:
            List of return values from all executed hooks
        """
        if event not in self.hooks:
            return []

        results = []
        for hook in self.hooks[event]:
            if not hook.enabled:
                continue

            # Check event type filter - this allows hooks to only process
            # specific types of events (e.g., only LMCAISEvent)
            if hook.event_types and "event_obj" in kwargs:
                event_obj = kwargs["event_obj"]
                if (
                    hasattr(event_obj, "event_type")
                    and event_obj.event_type not in hook.event_types
                ):
                    continue

            try:
                # Handle both async and sync callbacks transparently
                if asyncio.iscoroutinefunction(hook.callback):
                    result = await hook.callback(*args, **kwargs)
                else:
                    result = hook.callback(*args, **kwargs)
                results.append(result)
            except Exception as e:
                # Log error but don't stop processing - hooks should not
                # break the main tracing pipeline
                print(f"Hook {hook.name} failed: {e}")

        return results


# Default hooks for common use cases
def create_default_hooks() -> HookManager:
    """Create hook manager with default hooks.

    Sets up a basic set of hooks that provide common functionality:
    - Session start logging
    - Event validation
    - Automatic event enrichment

    Returns:
        HookManager with default hooks registered
    """
    manager = HookManager()

    # Example: Log session starts - useful for debugging and monitoring
    async def log_session_start(session_id: str, metadata: dict[str, Any]):
        import os

        if os.getenv("SYNTH_TRACE_VERBOSE", "0") in ("1", "true", "True"):
            print(f"Session started: {session_id}")

    # Example: Validate events before recording - ensures data quality
    def validate_event(event_obj: BaseEvent) -> bool:
        if not event_obj.system_instance_id:
            raise ValueError("Event must have system_instance_id")
        return True

    # Example: Add metadata to all events - useful for versioning and tracking
    def enrich_event(event_obj: BaseEvent):
        if "enriched" not in event_obj.metadata:
            event_obj.metadata["enriched"] = True
            event_obj.metadata["hook_version"] = "1.0"

    # Register default hooks
    manager.register("session_start", log_session_start, priority=10)
    manager.register("event_recorded", validate_event, name="validate_event", priority=20)
    manager.register("event_recorded", enrich_event, name="enrich_event", priority=15)

    return manager


# Global hooks instance - applications can customize this or create their own
# HookManager instances for different use cases
GLOBAL_HOOKS = create_default_hooks()

File: synth_ai_core/src/tracing/hooks.rs
//! Hook system for tracing events.
//!
//! Allows registering callbacks that are triggered at various points
//! during trace recording.

use std::collections::HashMap;
use std::sync::Arc;

use super::models::{MarkovBlanketMessage, TracingEvent};

/// Events that can trigger hooks.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum HookEvent {
    /// Session started
    SessionStart,
    /// Session ended
    SessionEnd,
    /// Timestep started
    TimestepStart,
    /// Timestep ended
    TimestepEnd,
    /// Event recorded
    EventRecorded,
    /// Message recorded
    MessageRecorded,
    /// Before saving to storage
    BeforeSave,
    /// After saving to storage
    AfterSave,
}

impl std::fmt::Display for HookEvent {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            HookEvent::SessionStart => write!(f, "session_start"),
            HookEvent::SessionEnd => write!(f, "session_end"),
            HookEvent::TimestepStart => write!(f, "timestep_start"),
            HookEvent::TimestepEnd => write!(f, "timestep_end"),
            HookEvent::EventRecorded => write!(f, "event_recorded"),
            HookEvent::MessageRecorded => write!(f, "message_recorded"),
            HookEvent::BeforeSave => write!(f, "before_save"),
            HookEvent::AfterSave => write!(f, "after_save"),
        }
    }
}

/// Context passed to hook callbacks.
#[derive(Debug, Clone, Default)]
pub struct HookContext {
    /// Current session ID (if any)
    pub session_id: Option<String>,
    /// Current step ID (if any)
    pub step_id: Option<String>,
    /// Event that triggered the hook (for EventRecorded)
    pub event: Option<TracingEvent>,
    /// Message that triggered the hook (for MessageRecorded)
    pub message: Option<MarkovBlanketMessage>,
}

impl HookContext {
    /// Create a new empty context.
    pub fn new() -> Self {
        Self::default()
    }

    /// Set the session ID.
    pub fn with_session(mut self, session_id: impl Into<String>) -> Self {
        self.session_id = Some(session_id.into());
        self
    }

    /// Set the step ID.
    pub fn with_step(mut self, step_id: impl Into<String>) -> Self {
        self.step_id = Some(step_id.into());
        self
    }

    /// Set the event.
    pub fn with_event(mut self, event: TracingEvent) -> Self {
        self.event = Some(event);
        self
    }

    /// Set the message.
    pub fn with_message(mut self, message: MarkovBlanketMessage) -> Self {
        self.message = Some(message);
        self
    }
}

/// Type alias for hook callbacks.
///
/// Callbacks receive a reference to the hook context and should not block.
pub type HookCallback = Arc<dyn Fn(&HookContext) + Send + Sync>;

/// A registered hook with its callback and priority.
struct HookRegistration {
    callback: HookCallback,
    priority: i32,
}

/// Manages hook registrations and triggering.
pub struct HookManager {
    hooks: HashMap<HookEvent, Vec<HookRegistration>>,
}

impl Default for HookManager {
    fn default() -> Self {
        Self::new()
    }
}

impl HookManager {
    /// Create a new hook manager.
    pub fn new() -> Self {
        Self {
            hooks: HashMap::new(),
        }
    }

    /// Register a hook callback for an event.
    ///
    /// Higher priority callbacks are executed first.
    pub fn register(&mut self, event: HookEvent, callback: HookCallback, priority: i32) {
        let registrations = self.hooks.entry(event).or_insert_with(Vec::new);

        registrations.push(HookRegistration { callback, priority });

        // Sort by priority descending (higher priority first)
        registrations.sort_by(|a, b| b.priority.cmp(&a.priority));
    }

    /// Register a hook with default priority (0).
    pub fn on(&mut self, event: HookEvent, callback: HookCallback) {
        self.register(event, callback, 0);
    }

    /// Trigger all hooks for an event.
    ///
    /// Hooks are called in priority order. Failures in hooks do not
    /// propagate - they are logged but don't stop other hooks.
    pub fn trigger(&self, event: HookEvent, context: &HookContext) {
        if let Some(registrations) = self.hooks.get(&event) {
            for reg in registrations {
                // Call the hook - we don't propagate panics
                std::panic::catch_unwind(std::panic::AssertUnwindSafe(|| {
                    (reg.callback)(context);
                }))
                .ok();
            }
        }
    }

    /// Check if any hooks are registered for an event.
    pub fn has_hooks(&self, event: HookEvent) -> bool {
        self.hooks
            .get(&event)
            .map(|v| !v.is_empty())
            .unwrap_or(false)
    }

    /// Get the number of hooks registered for an event.
    pub fn hook_count(&self, event: HookEvent) -> usize {
        self.hooks.get(&event).map(|v| v.len()).unwrap_or(0)
    }

    /// Clear all hooks for an event.
    pub fn clear(&mut self, event: HookEvent) {
        self.hooks.remove(&event);
    }

    /// Clear all hooks.
    pub fn clear_all(&mut self) {
        self.hooks.clear();
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::sync::atomic::{AtomicI32, Ordering};

    #[test]
    fn test_hook_registration() {
        let mut manager = HookManager::new();

        assert!(!manager.has_hooks(HookEvent::SessionStart));

        manager.on(HookEvent::SessionStart, Arc::new(|_| {}));

        assert!(manager.has_hooks(HookEvent::SessionStart));
        assert_eq!(manager.hook_count(HookEvent::SessionStart), 1);
    }

    #[test]
    fn test_hook_trigger() {
        let mut manager = HookManager::new();
        let counter = Arc::new(AtomicI32::new(0));
        let counter_clone = counter.clone();

        manager.on(
            HookEvent::SessionStart,
            Arc::new(move |_| {
                counter_clone.fetch_add(1, Ordering::SeqCst);
            }),
        );

        let context = HookContext::new().with_session("test-session");
        manager.trigger(HookEvent::SessionStart, &context);

        assert_eq!(counter.load(Ordering::SeqCst), 1);
    }

    #[test]
    fn test_hook_priority() {
        let mut manager = HookManager::new();
        let order = Arc::new(std::sync::Mutex::new(Vec::new()));

        let order1 = order.clone();
        manager.register(
            HookEvent::SessionStart,
            Arc::new(move |_| {
                order1.lock().unwrap().push(1);
            }),
            10,
        );

        let order2 = order.clone();
        manager.register(
            HookEvent::SessionStart,
            Arc::new(move |_| {
                order2.lock().unwrap().push(2);
            }),
            20, // Higher priority, should run first
        );

        let order3 = order.clone();
        manager.register(
            HookEvent::SessionStart,
            Arc::new(move |_| {
                order3.lock().unwrap().push(3);
            }),
            5,
        );

        manager.trigger(HookEvent::SessionStart, &HookContext::new());

        let result = order.lock().unwrap();
        assert_eq!(*result, vec![2, 1, 3]); // 20, 10, 5 priority order
    }

    #[test]
    fn test_hook_context() {
        let context = HookContext::new()
            .with_session("session-1")
            .with_step("step-1");

        assert_eq!(context.session_id, Some("session-1".to_string()));
        assert_eq!(context.step_id, Some("step-1".to_string()));
    }

    #[test]
    fn test_hook_clear() {
        let mut manager = HookManager::new();

        manager.on(HookEvent::SessionStart, Arc::new(|_| {}));
        manager.on(HookEvent::SessionEnd, Arc::new(|_| {}));

        assert!(manager.has_hooks(HookEvent::SessionStart));

        manager.clear(HookEvent::SessionStart);
        assert!(!manager.has_hooks(HookEvent::SessionStart));
        assert!(manager.has_hooks(HookEvent::SessionEnd));

        manager.clear_all();
        assert!(!manager.has_hooks(HookEvent::SessionEnd));
    }
}
